File: ios/Classes/Bridges/PaywallSkippedReasonBridge.swift
import Flutter
import SuperwallKit

public class PaywallSkippedReasonBridge: BridgeInstance {
  var reason: PaywallSkippedReason {
    assertionFailure("Subclasses must implement")
    return .placementNotFound
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "getDescription":
        let description = reason.description
        result(description)
      default:
        result(FlutterMethodNotImplemented)
    }
  }
}

public class PaywallSkippedReasonHoldoutBridge: PaywallSkippedReasonBridge {
  class override var bridgeClass: BridgeClass { "PaywallSkippedReasonHoldoutBridge" }
  override var reason: PaywallSkippedReason { internalReason }

  let internalReason: PaywallSkippedReason

  required init(bridgeId: BridgeId, initializationArgs: [String : Any]? = nil) {
    guard let reason = initializationArgs?["reason"] as? PaywallSkippedReason else {
      fatalError("Attempting to create a `PaywallSkippedReasonHoldoutBridge` without providing a reason.")
    }

    self.internalReason = reason
    super.init(bridgeId: bridgeId, initializationArgs: initializationArgs)
  }

  public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "getExperimentBridgeId":
        guard case .holdout(let experiment) = reason else {
          result(FlutterMethodNotImplemented)
          return
        }

        let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: ExperimentBridge.bridgeClass, initializationArgs: ["experiment": experiment])
        result(bridgeInstance.bridgeId)

      default:
        super.handle(call, result: result)
    }
  }
}

public class PaywallSkippedReasonNoAudienceMatchBridge: PaywallSkippedReasonBridge {
  class override var bridgeClass: BridgeClass { "PaywallSkippedReasonNoAudienceMatchBridge" }
  override var reason: PaywallSkippedReason { .noAudienceMatch }
}

public class PaywallSkippedReasonPlacementNotFoundBridge: PaywallSkippedReasonBridge {
  class override var bridgeClass: BridgeClass { "PaywallSkippedReasonPlacementNotFoundBridge" }
  override var reason: PaywallSkippedReason { .placementNotFound }
}

extension PaywallSkippedReason {
  func createBridgeId() -> BridgeId {
    switch self {
    case .holdout(_):
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: PaywallSkippedReasonHoldoutBridge.bridgeClass, initializationArgs: ["reason": self])
      return bridgeInstance.bridgeId

    case .noAudienceMatch:
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: PaywallSkippedReasonNoAudienceMatchBridge.bridgeClass, initializationArgs: ["reason": self])
      return bridgeInstance.bridgeId

    case .placementNotFound:
      let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: PaywallSkippedReasonPlacementNotFoundBridge.bridgeClass, initializationArgs: ["reason": self])
      return bridgeInstance.bridgeId
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/PaywallSkippedReasonBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import com.superwall.sdk.paywall.presentation.internal.state.PaywallSkippedReason
import com.superwall.superwallkit_flutter.BridgingCreator
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch

abstract class PaywallSkippedReasonBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : BridgeInstance(context, bridgeId, initializationArgs), MethodChannel.MethodCallHandler {
    abstract val reason: PaywallSkippedReason

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getDescription" -> {
                val description = reason.toString()
                result.success(description)
            }

            else -> result.notImplemented()
        }
    }
}

class PaywallSkippedReasonHoldoutBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PaywallSkippedReasonBridge(context, bridgeId, initializationArgs) {
    private val scope = CoroutineScope(Dispatchers.IO)

    companion object {
        fun bridgeClass(): BridgeClass = "PaywallSkippedReasonHoldoutBridge"
    }

    override val reason: PaywallSkippedReason =
        initializationArgs?.get("reason") as? PaywallSkippedReason
            ?: throw IllegalArgumentException("Attempting to create a `PaywallSkippedReasonHoldoutBridge` without providing a reason.")

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {

        when (call.method) {
            "getExperimentBridgeId" -> {
                scope.launch {
                    if (reason is PaywallSkippedReason.Holdout) {
                        val experiment = reason.experiment
                        val experimentBridgeId =
                            BridgingCreator.shared.createBridgeInstanceFromBridgeClass(
                                ExperimentBridge.bridgeClass(),
                                mapOf("experiment" to experiment)
                            )
                        result.success(experimentBridgeId)
                    } else {
                        result.notImplemented()
                    }
                }
            }

            else -> super.onMethodCall(call, result)
        }
    }
}

class PaywallSkippedReasonNoAudienceMatchBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PaywallSkippedReasonBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PaywallSkippedReasonNoAudienceMatchBridge" }
    override val reason: PaywallSkippedReason = PaywallSkippedReason.NoAudienceMatch()
}

class PaywallSkippedReasonPlacementNotFoundBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PaywallSkippedReasonBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PaywallSkippedReasonPlacementNotFoundBridge" }
    override val reason: PaywallSkippedReason = PaywallSkippedReason.PlacementNotFound()
}

suspend fun PaywallSkippedReason.createBridgeId(): BridgeId {
    return when (this) {
        is PaywallSkippedReason.Holdout -> {
            val bridgeInstance = BridgingCreator.shared.createBridgeInstanceFromBridgeClass(
                PaywallSkippedReasonHoldoutBridge.bridgeClass(),
                mapOf("reason" to this)
            )
            return bridgeInstance.bridgeId
        }

        is PaywallSkippedReason.NoAudienceMatch -> {
            val bridgeInstance = BridgingCreator.shared.createBridgeInstanceFromBridgeClass(
                PaywallSkippedReasonNoAudienceMatchBridge.bridgeClass(),
                mapOf("reason" to this)
            )
            return bridgeInstance.bridgeId
        }

        is PaywallSkippedReason.PlacementNotFound -> {
            val bridgeInstance = BridgingCreator.shared.createBridgeInstanceFromBridgeClass(
                PaywallSkippedReasonPlacementNotFoundBridge.bridgeClass(),
                mapOf("reason" to this)
            )
            return bridgeInstance.bridgeId
        }
    }
}
